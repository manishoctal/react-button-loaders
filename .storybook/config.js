import { configure } from '@storybook/react';
import { injectGlobal } from 'styled-components';

injectGlobal`
  @import url('https://fonts.googleapis.com/css?family=Heebo:400,600');
  button span {
    font-family: 'Heebo', sans-serif;
    font-weight: 600;
  }
`
// automatically import all files ending in *.stories.js
const req = require.context('../src/stories', true, /.stories.js$/);
function loadStories() {
  req.keys().forEach((filename) => req(filename));
}

configure(loadStories, module);
