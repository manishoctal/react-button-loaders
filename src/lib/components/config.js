export default {
  prefixCls: 'react-button-loaders'
}
