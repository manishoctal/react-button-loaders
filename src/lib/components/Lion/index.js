import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import config from '../../config.js'
const { prefixCls } = config
import {
  ButtonStyled,
  Text,
  Loading,
  Success,
  RefreshIcon,
  CheckIcon,
  Effect
} from './style'

const Lion = ({
  children,
  className,
  state,
  refreshComponent,
  checkComponent,
  textWhileLoading,
  textWhenLoaded,
  ...props
}) => {
  const classNameJoined = classNames(className, `${prefixCls}__btn`, {
    [`${prefixCls}__loading`]: state === 'loading',
    [`${prefixCls}__finished`]: state === 'finished'
  })
  return (
    <ButtonStyled {...props} className={classNameJoined}>
      <Text>{children}</Text>
      <Effect />
      <Loading>
        <span>{textWhileLoading}</span>
      </Loading>
      <Success>
        {checkComponent}
        <span>{textWhenLoaded}</span>
      </Success>
    </ButtonStyled>
  )
}

Lion.defaultProps = {
  className: '',
  speedProgress: 3000,
  speedIconLoader: 800,
  refreshComponent: <RefreshIcon />,
  checkComponent: <CheckIcon />,
  onClick: () => {},
  bgColor: '#F06189',
  bgLoading: '#F06189',
  bgWhenFinish: '#4AD481',
  bgLoadingBehind: '#D7E3EF'
}

const { string, number, func } = PropTypes

Lion.propTypes = {
  textWhenLoaded: string,
  textWhileLoading: string,
  className: string,
  speedProgress: number,
  speedIconLoader: number,
  onClick: func,
  bgColor: string,
  bgLoading: string,
  bgLoadingBehind: string,
  bgWhenFinish: string
}

export default Lion
