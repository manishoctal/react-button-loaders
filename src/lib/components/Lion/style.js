import React from 'react'
import styled, { keyframes } from 'styled-components'
import { IconRefresh, IconCheck } from '../Icons'
import {
  rotate360,
  progressBar,
  scaleFinished,
  toWidth100,
  toHeight100
} from '../../animations'

import config from '../../config'
const { prefixCls } = config

export const Text = styled.span`
  font-size: 15px;
`
Text.displayName = 'Text'

export const Loading = styled.span``
Loading.displayName = 'Loading'

export const Success = styled.span``
Success.displayName = 'Success'

const RefreshStyled = styled.span`
  svg {
    fill: #fff;
    width: 20px;
  }
`

const CheckStyled = styled.span`
  svg {
    stroke: #FFF
    width: 24px;
    height: 24px;
  }
`

export const Effect = styled.span``

export const RefreshIcon = () => (
  <RefreshStyled>
    <IconRefresh />
  </RefreshStyled>
)

export const CheckIcon = () => (
  <CheckStyled>
    <IconCheck />
  </CheckStyled>
)

export const ButtonStyled = styled.button`
  position: relative;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  width: 200px;
  padding: 24px 50px;
  background-color: ${props => props.bgColor};
  color: #fff;
  border: none;
  border-radius: 4px;
  text-transform: uppercase;
  overflow: hidden;
  outline: none;
  transition: all 0.2s;

  &:hover {
    box-shadow: 0 12px 24px 0 rgba(233, 92, 97, 0.2);
  }

  &:before {
    position: absolute;
    content: '';
    bottom: 0;
    left: 0;
    transform: translateX(-100%);
    width: 100%;
    height: 100%;
    background-color: ${props => props.bgLoading};
  }

  ${Text}, ${Loading}, ${Success} {
    position: absolute;
    line-height: 0;
  }
  ${Text} {
    top: 50%;
    transform: translateY(-50%);
  }
  ${Loading} {
    top: 200%;
    transform: translateY(0%);
  }
  ${Success} {
    display: none;
  }
  &.${prefixCls}__loading {
    padding: 0;
    height: 20px;
    transform: translateY(15px);
    cursor: progress;
    pointer-events: none;
    background-color: ${props => props.bgLoadingBehind};
    &:before {
      animation: ${progressBar} ${props => props.speedProgress}ms linear
        infinite;
    }

    & > span {
      &:nth-of-type(1) {
        top: -100%;
        transform: translateY(-50%);
      }

      &:nth-of-type(3) {
          top: 50%;
      }
    }
  }
  &.${prefixCls}__finished {
    background: #FFF;
    &:before, &:after {
        content:"";
        width: 0;
        height: 2px;
        position: absolute;
        background: ${props => props.bgWhenFinish}
        left: initial;
        bottom: initial;
        transform: initial;
    }
    &:before {
        right: 0;
        top: 0;
    }
    &:after {
        left: 0;
        bottom: 0;
    }
    ${Effect}{
        &:before, &:after{
            content:"";
            width:2px;
            height:0;
            position: absolute;
            transition: all 0.1s linear;
            background: ${props => props.bgWhenFinish}
        }
        &:before {
            left: 0;
            top: 0;
        }
        &:after {
            right: 0;
            bottom: 0
        }
    }
    ${Success} {
        &:after {
            transition-delay: 0.2s;
        }
        &:before {
            transition-delay: 0.6s;
        }
    }
    &:before {
        animation: ${toWidth100} .1s linear;
        animation-fill-mode: forwards;
        animation-delay: .4s;
    }
    &:after {
        animation: ${toWidth100} .1s linear;
        animation-fill-mode: forwards;
        animation-delay: 0s;
    }
    ${Effect} {
        &:before {
            animation: ${toHeight100} .1s linear;
            animation-fill-mode: forwards;
            animation-delay: .6s;
        }
        &:after {
            animation: ${toHeight100} .1s linear;
            animation-fill-mode: forwards;
            animation-delay: .2s;
        }
    }
    ${Text}, ${Loading} {
      display: none;
    }
    ${Success} {
      display: flex !important;
      align-items: center;
      animation: ${scaleFinished} 0.5s linear;
      #icon-check {
        stroke: ${props => props.bgWhenFinish};
      }
      & > span {
        margin-left: 5px;
        font-size: 12px;
        color: ${props => props.bgWhenFinish};
      }
    }
  }
`
