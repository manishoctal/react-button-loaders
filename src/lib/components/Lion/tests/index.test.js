import React from 'react'
import config from '../../../config.js'
const { prefixCls } = config
import Lion from '../index'
import { shallow } from 'enzyme'

describe('Lion component', () => {
  test('should render a the button', () => {
    const wrapper = shallow(<Lion />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should call mock function when button is clicked', () => {
    const handleClick = jest.fn()
    const wrapper = shallow(<Lion onClick={handleClick} />)
    wrapper.simulate('click')
    expect(handleClick).toHaveBeenCalled()
  })

  test('should put class `__loading` when the state is loading ', () => {
    const wrapper = shallow(<Lion state='loading' />)
    expect(wrapper.find(`.${prefixCls}__loading`).length).toEqual(1)
  })

  test('should put class `__finished` when the state is finished ', () => {
    const wrapper = shallow(<Lion state='finished' />)
    expect(wrapper.find(`.${prefixCls}__finished`).length).toEqual(1)
  })
})
