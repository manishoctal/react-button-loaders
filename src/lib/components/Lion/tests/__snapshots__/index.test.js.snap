// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lion component should render a the button 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Lion
    bgColor="#F06189"
    bgLoading="#F06189"
    bgLoadingBehind="#D7E3EF"
    bgWhenFinish="#4AD481"
    checkComponent={<Unknown />}
    className=""
    onClick={[Function]}
    refreshComponent={<Unknown />}
    speedIconLoader={800}
    speedProgress={3000}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "bgColor": "#F06189",
      "bgLoading": "#F06189",
      "bgLoadingBehind": "#D7E3EF",
      "bgWhenFinish": "#4AD481",
      "children": Array [
        <Text />,
        <styled.span />,
        <Loading>
          <span />
</Loading>,
        <Success>
          <Unknown />
          <span />
</Success>,
      ],
      "className": "react-button-loaders__btn",
      "onClick": [Function],
      "speedIconLoader": 800,
      "speedProgress": 3000,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <span />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "span",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Unknown />,
            <span />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "span",
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "bgColor": "#F06189",
        "bgLoading": "#F06189",
        "bgLoadingBehind": "#D7E3EF",
        "bgWhenFinish": "#4AD481",
        "children": Array [
          <Text />,
          <styled.span />,
          <Loading>
            <span />
</Loading>,
          <Success>
            <Unknown />
            <span />
</Success>,
        ],
        "className": "react-button-loaders__btn",
        "onClick": [Function],
        "speedIconLoader": 800,
        "speedProgress": 3000,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <span />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "span",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Unknown />,
              <span />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
              },
              "ref": null,
              "rendered": null,
              "type": "span",
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
